### PR TITLE
feat(edgeless): support edgeless mode create linked doc

### PR DESCRIPTION
**Pull request by akumatus**
### TL;DR
This pull request support edgeless mode to create linked doc. Find more details in [THIS RFC](https://app.affine.pro/workspace/af3478a2-9c9c-4d16-864d-bffa1eb10eb6/XUmNieqGMKDs525xNqY8G).

There are two way to create a linked doc in edgeless editor:
- The first way, select a single note and click the `Turn into linked doc` menu item. The content of the selected note will become a independant document.

<div class='graphite__hidden'>
          <div>🎥 Video uploaded on Graphite:</div>
            <a href="https://app.graphite.dev/media/video/sJGviKxfE3Ap685cl5bj/029fa907-26a5-4063-9fb2-0409931ab291.mov">
              
            </a>
          </div>
<video src="https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/sJGviKxfE3Ap685cl5bj/029fa907-26a5-4063-9fb2-0409931ab291.mov">录屏2024-06-03 18.52.41.mov</video>

- The second way, select multiple edgeless elements and click the `create linked doc` menu item. All the selected elements will be moved to the linked document.

<div class='graphite__hidden'>
          <div>🎥 Video uploaded on Graphite:</div>
            <a href="https://app.graphite.dev/media/video/sJGviKxfE3Ap685cl5bj/06852f1b-7b16-4745-a8d8-24f1770ee306.mov">
              
            </a>
          </div>
<video src="https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/sJGviKxfE3Ap685cl5bj/06852f1b-7b16-4745-a8d8-24f1770ee306.mov">录屏2024-06-03 18.50.04.mov</video>

### What changed?
Features:
- Add `Turn into linked doc` and `Create linked doc` item in more menu, as well as handlers for the menu items.
- Add three kind of linked doc create method:
  - createLinkedDocFromBlocks
  - createLinkedDocFromNote
  - createLinkedDocFromEdgelessElements
- Add `sortEdgelessElements` to sort elements before added to edgeless root element.
- Add`mapConnectorIds` to maintain interdependencies between nodes. 
Since the id of the element in linked doc will be regenerated，using `mapConnectorIds` to complete the mapping of old and new element ids. Three kind of interdependencies:
  - source and target of connector
  - children of groups
  - parent and children of mindmaps
- Add e2e tests for edgeless mode create linked doc scenes.
- Add `SerializedElement` typescript type for better type inference.

Refactor:
- Move `getBlockProps` function from embed card popover to `common/utils/block-props.ts`, thus it can be easily reused.
- Extract side effects of `createLinkedDocFromSelectedBlocks`, using the first heading block content as doc title only works in page mode.
- Split the move blocks into two parts: add blocks and delete blocks, so they can be easily reused.

### How to test?

As shown in the screen recording video before, select any elements on the edgeless board and convert it in to linked doc. Get more inspiration based on scenarios from e2e testing.

---

### CODE REVIEW

**Review by Saul-Mirone — APPROVED** [review]

File: packages/blocks/src/_common/utils/render-linked-doc.ts [hunk 1]
```diff
@@ -346,79 +354,151 @@ async function renderImageAbstract(
   card.isBannerEmpty = false;
 }
 
-function moveBlocksToLinkedDoc(
-  doc: Doc,
-  linkedDoc: Doc,
+export function addBlocksToNote(
```
Comment (doouding) [inline packages/blocks/src/_common/utils/render-linked-doc.ts:357]: Not sure if the name should be `addBlocksToDoc` ?

File: packages/blocks/src/_common/utils/render-linked-doc.ts [hunk 2]
```diff
@@ -346,79 +354,151 @@ async function renderImageAbstract(
...
   model: BlockModel,
   parentId: string
 ) {
   // Add current block to linked doc
-  const keys = model.keys as (keyof typeof model)[];
-  const values = keys.map(key => model[key]);
-  const blockProps = Object.fromEntries(keys.map((key, i) => [key, values[i]]));
-  const newModelId = linkedDoc.addBlock(
-    model.flavour as never,
+  const blockProps = getBlockProps(model);
```
Comment (doouding) [inline packages/blocks/src/_common/utils/render-linked-doc.ts:363]: There might also be some blocks' props happend to be another blocks's id that need to be remapped. Eg. `affine:surface-ref`.

File: packages/blocks/src/_common/utils/render-linked-doc.ts (hunk 1 above)
Comment (akumatus) [reply packages/blocks/src/_common/utils/render-linked-doc.ts]: Make sense. I have changed the function name, thanks for your comment!

File: packages/blocks/src/_common/utils/render-linked-doc.ts (hunk 2 above)
Comment (akumatus) [reply packages/blocks/src/_common/utils/render-linked-doc.ts:363]: Got it. I will open another PR to fix this case specifically.

Comment (akumatus) [reply packages/blocks/src/_common/utils/render-linked-doc.ts:363]: https://github.com/toeverything/blocksuite/assets/12724894/4ba8475f-b0f7-4ec2-b9e5-e16c685bc43c
